internal/config/config.go

```diff
@@ -14,23 +17,50 @@ type ConfigurationValidator interface {
 type Configuration struct {
 	Debug      bool
 	Verbosity  int
-	SharedPath string
+	SharedPath string `validate:"required"`
 	Temporal   Temporal
 	Worker     WorkerConfig
 }
 
 type Temporal struct {
-	Address      string
-	Namespace    string
-	TaskQueue    string
-	WorkflowName string
+	// Address is the Temporal server host and port (default: "localhost:7233").
+	Address string
+
+	// Namespace is the Temporal namespace the preprocessing worker should run
+	// in (default: "default").
+	Namespace string
+
+	// TaskQueue is the Temporal task queue from which the preprocessing worker
+	// will pull tasks.
+	TaskQueue string `validate:"required"`
+
+	// WorkflowName is the name of the preprocessing Temporal workflow.
+	WorkflowName string `validate:"required"`
 }
 
 type WorkerConfig struct {
-	MaxConcurrentSessions int
+	// MaxConcurrentSessions limits the number of workflow sessions the
+	// preprocessing worker can handle simultaneously (default: 1).
+	MaxConcurrentSessions int `validate:"gte=1"`
 }
 
-func (c Configuration) Validate() error { return nil }
+func (c Configuration) Validate() error {
+	var errs error
+
+	validate := validator.New(validator.WithRequiredStructEnabled())
+	if err := validate.Struct(c); err != nil {
```

Comment:
This is going to validate the inner `bagit` config struct directly here when we add it, right? That will make the `Validate` function kind of a duplication there and, with the error formatting, we'll need to know about the possible errors the inner struct may trigger. But we may still want the `Validate` function in the activity config for cases where it's not configurable through the worker config and needs to be validated independently.

While I like this validation idea, I don't like how it cascades considering that config structs are in different projects. I'd rather see independent `Validate` functions that could be called from the worker config validation and from other places.

Reply:
@jraddaoui yes, the `validate.Struct(c)` method would cascade the validation to embedded structs like `bagit.Config`.  My plan was to eliminate the `bagit.Config.Validate()` method altogether, and rely on the cascading feature to validate against the struct field "validate" tags.  I would move the validation error formatting for the "oneof" tag to this method, so it would consistently format any "oneof" messages from any embedded struct.

I think there are multiple trade-offs to using a single, central validator like this vs. distributing validation responsibilities to the imported packages and at this point I'm on the fence about whether to use `github.com/go-playground/validator` or not. So I think it's time for a pros/cons list!

PROS of github.com/go-playground/validator: 

- Struct tags are a elegant way to specify validation constraints - they are specified locally but can be validated globally, and they can be totally ignored if validation isn't required
- We don't have to explicitly call a `Validate()` method for each imported config that we want to validate
- The "oneof" struct tag is especially nice because it allows specifying a controlled list of values without having to create a Go pseudo-enum
- Caching of validation constraints and evaluation (but this probably doesn't matter if we only call `validate.Struct()` once)

CONS:

- Probably overkill. We don't need a lot of the features and it's a pretty big third-party dependency for our project.
- Validation struct tags are a bit "magical" - it may not be clear how validation works to developers that aren't familiar with Go struct tags
- Struct tags could get long and hard to read if there are tags for Marshalling/Unmarshalling and validation
- The validation error messages are not clear, and it's a bit cumbersome to improve them
- Centralized validation instead of distributed?  (I don't actually see a problem with this, but maybe there's something I'm missing?)

@jraddaoui do you have Pros or Cons to add, or want to debate any of the above?  Maybe this is a good discussion to have in a video chat because I can imagine a lot of back and forth discussion.

Reply:
Thanks for putting this together @djjuhasz, lets talk about it after the standup if you have some time ;)

Reply:
I've removed github.com/go-playground/validator in favour of a simple `Validate()` method with no dependencies.